sympy/stats/stochastic_process_types.py

```diff
@@ -224,12 +234,17 @@ def joint_distribution(self, *args):
                 raise ValueError("Expected a RandomIndexedSymbol or "
                                 "key not  %s"%(type(arg)))
 
-        if args[0].pspace.distribution == None: # checks if there is any distribution available
+        t = Dummy('t')
+        dist = self.distribution(t)
+        if type(dist) is Distribution: # checks if there is any distribution available
             return JointDistribution(*args)
-
-        pdf = Lambda(tuple(args),
-                expr=Mul.fromiter(arg.pspace.process.density(arg) for arg in args))
-        return JointDistributionHandmade(pdf)
+        if hasattr(dist, 'pdf'):
+            density = Lambda(tuple(args),
+                    expr=Mul.fromiter(self.distribution(arg.key).pdf(arg) for arg in args))
+        elif hasattr(dist, 'pmf'):
+            density = Lambda(tuple(args),
+                    expr=Mul.fromiter(self.distribution(arg.key).pmf(arg) for arg in args))
+        return JointDistributionHandmade(density)
```

Comment:
```suggestion
        elif hasattr(dist, 'pmf'):
            density = Lambda(tuple(args),
                    expr=Mul.fromiter(self.distribution(arg.key).pmf(arg) for arg in args))
        if density is None:
            return JointDistribution(density)
        return JointDistributionHandmade(density)
```

Reply:
Finally, we still have to check `if density is None`, so can you please elaborate on what is the advantage of changing the current implementation to your suggestions? In fact, doesn't the current implementation provide an early exit if there is no implementation, rather than having to go through the two `if`s, so isn't it better? At least that's the ideology I have been following so far.

Reply:
Well the other test is a bit cryptic and heavily dependent on python and whenever we can avoid it we should. There are other places where these kind of type checks are there and if possible the code should be refactored to avoid those type checks if possible.

>  In fact, doesn't the current implementation provide an early exit

Don't worry too much about micro-optimisations. 

P.S. That `hasattr` thing also seems cryptic. May be in a different PR that should be replaced with a not so pythonic check.